java/ql/src/experimental/Security/CWE/CWE-470/LoadClassNoSignatureCheck.ql

```diff
@@ -0,0 +1,111 @@
+            getControlFlowNodeSuccessor+(ma.getControlFlowNode()) = maInvoke and
+            maInvoke.getCallee().hasName("invoke") and
+            result = maInvoke
+            )
+    ) or
+    (
+        exists(AssignExpr ae, VarAccess va1, VarAccess va2, MethodAccess maInvoke |
+            ae.getSource() = ma and
+            ae.getDest() = va1 and
+            maInvoke.getQualifier() = va2 and
+            va1.getVariable() = va2.getVariable() and
+            result = maInvoke
+            )
+    ))
+}
+
+module SignaturePackageConfig implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) {
+        exists(MethodAccess maCheckSignatures |
+            maCheckSignatures.getCallee().getDeclaringType().getQualifiedName() = "android.content.pm.PackageManager" and
+            maCheckSignatures.getCallee().getName() = "checkSignatures" and
+            source.asExpr() = maCheckSignatures.getArgument(0)
+            )
+    }
+
+    predicate isSink(DataFlow::Node sink) {
+        exists (MethodAccess maCreatePackageContext |
+            (maCreatePackageContext.getCallee().getDeclaringType().getQualifiedName() = "android.content.ContextWrapper" or
+             maCreatePackageContext.getCallee().getDeclaringType().getQualifiedName() = "android.content.Context") and 
+            maCreatePackageContext.getCallee().getName() = "createPackageContext" and
+            sink.asExpr() = maCreatePackageContext.getArgument(0)
+            )                
+    }
+}
+
+module SigPkgCfg = TaintTracking::Global<SignaturePackageConfig>;
+
+predicate isSignaturesChecked(MethodAccess maCreatePackageContext)
+{
+    exists(DataFlow::Node source, DataFlow::Node sink |
+        SigPkgCfg::flow(source, sink) and
+        sink.asExpr() = maCreatePackageContext.getArgument(0)
+        )
+}
+ 
+from 
+    MethodAccess maCreatePackageContext, 
+    LocalVariableDeclExpr lvdePackageContext,
+    DataFlow::Node sinkPackageContext,
+    MethodAccess maGetMethod,
+    MethodAccess maInvoke 
+where
+    (maCreatePackageContext.getCallee().getDeclaringType().getQualifiedName() = "android.content.ContextWrapper" or
+     maCreatePackageContext.getCallee().getDeclaringType().getQualifiedName() = "android.content.Context") and 
+    maCreatePackageContext.getCallee().getName() = "createPackageContext" and
+    not isSignaturesChecked(maCreatePackageContext) and
+    lvdePackageContext.getEnclosingStmt() = maCreatePackageContext.getEnclosingStmt() and
+    TaintTracking::localTaint(DataFlow::exprNode(lvdePackageContext.getAnAccess()), sinkPackageContext) and
+    getClassLoaderReachableMethodAccess(sinkPackageContext) = maGetMethod and
+    getDangerousReachableMethodAccess(maGetMethod) = maInvoke
+select
+    lvdePackageContext, 
+    sinkPackageContext, 
+    maGetMethod,
+    maInvoke,
+    "Potential arbitary code execution due to class loading without package signature checking."
```

Comment:
Recommend some adjustments to the select clause so that it's simpler and it works well with a `@kind problem`:

```suggestion
from
  MethodCall maCreatePackageContext, LocalVariableDeclExpr lvdePackageContext,
  Expr sinkPackageContext, MethodCall maGetMethod, MethodCall maInvoke
where
  maCreatePackageContext
      .getMethod()
      .hasQualifiedName("android.content", ["ContextWrapper", "Context"], "createPackageContext") and
  not isSignaturesChecked(maCreatePackageContext) and
  lvdePackageContext.getEnclosingStmt() = maCreatePackageContext.getEnclosingStmt() and
  TaintTracking::localExprTaint(lvdePackageContext.getAnAccess(), sinkPackageContext) and
  getClassLoaderReachableMethodCall(sinkPackageContext) = maGetMethod and
  getGetMethodMethodCall(maGetMethod) = maInvoke
select maInvoke, "Potential arbitary code execution due to $@ without $@ signature checking.", sinkPackageContext, "class loading", sinkPackageContext, "package"
```